### DISCUSSION TITLE
System architecture

### DISCUSSION THREAD

**Discussion by dsychin** [B-Open/jobalert#2] [General]
Initial high level design, to be refined.

- Front facing website (accessing data via API or direct database connection)
- Database
- Background processes
  - Website scraper (write into database)
  - Alert sender

**Comment by dsychin** [comment]
Technology choices

Requirements:
- Should be able to run on a linux machine

- Database: MySQL (or others?)
- Background processes: C# .NET Core
- Front facing website: Django or .NET?
  - Do we need to use a front end js framework?

**Reply by syahnur197** [reply]
to make it easy to manage the system, we might as well make it monolithic application, let's have the frontend also handled by .NET

as for the Database, MySQL should be enough

**Reply by dsychin** [reply]
@syahnur197 What about the front end? Do we need a single page application?

**Reply by dsychin** [reply]
I would keep at least the web and background process part separate.

They can be in the same solution but it's probably a better idea to have them be able to run independent on each other.

**Reply by syahnur197** [reply]
don't you think it's a bit overengineering? But I'm fine with separate architecture as well.

In that case, I suggest we go with Next.JS, we can host the frontend for free at vercel

**Reply by dsychin** [reply]
At the very least I would want to keep those 2 separate since I had the issue with the background processes clashing with the web server.

It also means that we need to stop the web server if we want to make any changes to the background process and vice versa.

NextJS sounds fine, I've used it before. That would mean that the .NET web server will be an API server then.

**Reply by dsychin** [reply]
Overall it would look something like this.

![job alert drawio](https://user-images.githubusercontent.com/19268668/138395862-00e06844-4161-40b6-80fb-0518a8d2069b.png)

**Reply by syahnur197** [reply]
that means we would have 3 apps for this project. That would be fine by me. Though I need a bit of holding hand to code the .NET part. I have built an API using .NET before accessing MSSQL, I think the process flow should be the same for MySQL.

**Reply by dsychin** [reply]
Yes, I assumed you probably used Entity Framework for working with MSSQL?

We'll probably have to use something like Dapper instead which lets you write SQL code that maps to a class.

I have tried to use Entity Framework with PostgreSQL before and it was quite a nightmare to get things to work. ER works very well for MSSQL but for everything else it is a bit lacking. MSSQL is great but it costs money for anything more than a simple project haha

I can set up the .NET project first then we can build off from there. You can probably set up the front end first with static mock data.

**Reply by syahnur197** [reply]
I've heard of dapper as well, though I never had the opportunity to try it out. But I guess the theory would be the same.

Yes I agree with you managing MSSQL is tedious and I'm no that well verse with it.

Okay I will set up the front end part of the application

**Reply by dsychin** [reply]
@syahnur197 I have added some boilerplate .NET code.

btw, in case you're not aware, Hacktoberfest is currently on going until the end of October. If you make 4 qualifying pull requests, you get a free T-shirt.

https://hacktoberfest.digitalocean.com/

**Reply by syahnur197** [reply]
anw, if it's not too much to ask, is it okay if you provide documentation on the dotnet side's installation

**Reply by dsychin** [reply]
Yeah, I was planning to add it to the readme.

It is fairly simple, just install .NET 5 according to the official instructions, change directory into the folders inside src and call dotnet run or build.

```
cd src/Api
dotnet run
```

Then view the api at http://localhost:5000/weatherforecast

The Worker folder is just a hello world console app, while the Shared folder doesn't actually run.

The sql folder is using go-migrate as a cli tool to handle database migration. Currently just blank files.

May I ask what OS is your development environment on? My instructions are for Linux but it would be the same for Windows.

Alternatively you can just use Visual Studio and open the sln file to run it.

**Reply by syahnur197** [reply]
I'm using Pop OS (Linux) as well, hahaha that would be easier for both of us, I can just follow your instruction to the T

Since Migration is handled using go, I need to install go and also learn go to create db migration files correct?

What libraries are you using for the migration

**Reply by dsychin** [reply]
No need to install go, just install migrate as a binary.

https://github.com/golang-migrate/migrate/tree/master/cmd/migrate

I'm new to this tool but I've used a similar migration CLI that is based on Rust.

The advantage of this go based one is that you can just install it as binary instead of needing to install some other build tools.

I have only used 1 command to generate the initial files so far.

```bash
cd sql
migrate create -ext=sql init
```

**Reply by dsychin** [reply]
I'll probably add a bash script for regular migration tasks.

**Reply by dsychin** [reply]
@syahnur197 Can you add what you're working on into issues and assign yourself? That way we can keep track of what we are working on.

And also maybe work on a different branch and use pull requests so that we can review each other's work?

**Reply by syahnur197** [reply]
okay noted, I never used that approach before, so let me try

**Reply by dsychin** [reply]
@syahnur197 You can read more about it. It's called the feature branch workflow.

https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

There is also the Gitflow workflow, which is what is commonly used for open source development. The difference is that with gitflow you fork the main repo and create a branch to make your changes, instead of just creating a branch.

https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow
